### PR TITLE
2997: Rate Module

### LINKED ISSUE #2997: Creation of sub rates
related to enhancement mentioned below,
https://issues.apache.org/jira/browse/FINERACT-614

### PR DESCRIPTION
**Pull request by angelboxes**
New pages were added for Rate definition and rates can be added in loan products and accounts

## Description
A new rate module was added to create previously defined rates and use them to configure new loan accounts and loan products.

## Related issues and discussion
2997

## Screenshots, if any

screenshot_2018-08-08 mifos x client 1

screenshot_2018-08-08 mifos x client

screenshot_2018-08-08 mifos x client 2

## Checklist
Please make sure these boxes are checked before submitting your pull request - thanks!

- [x] Validate the JS and HTML files with `grunt validate` to detect errors and potential problems in JavaScript code.

- [x] Run the tests by opening `test/SpecRunner.html` in the browser to make sure you didn't break anything.

- [x] If you have multiple commits please combine them into one commit by squashing them.

- [x] Read and understood the contribution guidelines at `community-app/Contributing.md`.

### CODE REVIEW

File: app/scripts/controllers/loanAccount/EditLoanAccAppController.js [hunk 1]
```diff
@@ -162,7 +162,72 @@
...
+            //Rate
+            scope.rateSelected = function(currentRate){
+
+                if(currentRate && !scope.checkIfRateAlreadyExist(currentRate)){
+                    scope.rateFlag=true;
+                    scope.formData.rates.push(currentRate);
+                    scope.rateOptions.splice(scope.rateOptions.indexOf(currentRate),1);
+                    scope.currentRate = '';
+                    currentRate = '';
+                    console.log(scope.rateOptions);
```
Comment (ShruthiRajaram) [inline app/scripts/controllers/loanAccount/EditLoanAccAppController.js]: remove console.log

File: app/scripts/controllers/loanAccount/EditLoanAccAppController.js [hunk 2]
```diff
@@ -162,7 +162,72 @@
...
+
+                return exist;
+            };
+
+            scope.calculateRates = function(){
+                var total = 0;
+                scope.formData.rates.forEach(function(rate){
+                    total += rate.percentage;
+                });
+                if (total===0){
```
Comment (ShruthiRajaram) [inline app/scripts/controllers/loanAccount/EditLoanAccAppController.js:213]: is total not a numerical in any case? why === is used

File: app/scripts/controllers/loanAccount/NewLoanAccAppController.js [hunk 3]
```diff
@@ -177,9 +179,59 @@
...
+          //Rate
+          scope.rateSelected = function(currentRate){
+
+            if(currentRate && !scope.checkIfRateAlreadyExist(currentRate)){
+                scope.rateFlag=true;
+              scope.formData.rates.push(currentRate);
+              scope.rateOptions.splice(scope.rateOptions.indexOf(currentRate),1);
+              scope.currentRate = '';
+              currentRate = '';
+              console.log(scope.rateOptions);
```
Comment (ShruthiRajaram) [inline app/scripts/controllers/loanAccount/NewLoanAccAppController.js]: remove console log

File: app/scripts/controllers/product/CreateLoanProductController.js [hunk 4]
```diff
@@ -121,7 +128,53 @@
...
+              scope.rateFlag = false;
+            }
+
+            scope.formData.minInterestRatePerPeriod = minRate;
+            //Assign the same total range to this values.
+            scope.formData.interestRatePerPeriod = total;
+            scope.formData.maxInterestRatePerPeriod = total;
+            scope.calculatedRatePerPeriod = total;
+
+            console.log(scope.formData);
```
Comment (ShruthiRajaram) [inline app/scripts/controllers/product/CreateLoanProductController.js]: remove console log

File: app/scripts/controllers/product/CreateRateController.js [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+            // resourceFactory.chargeTemplateResource.get(function (data) {
+            //     scope.template = data;
+            //     scope.showChargePaymentByField = true;
+            //     scope.chargeCalculationTypeOptions = data.chargeCalculationTypeOptions;
+            //     scope.chargeTimeTypeOptions = data.chargeTimeTypeOptions;
+            //
+            //     scope.incomeAccountOptions = data.incomeOrLiabilityAccountOptions.incomeAccountOptions || [];
+            //     scope.liabilityAccountOptions = data.incomeOrLiabilityAccountOptions.liabilityAccountOptions || [];
+            //     scope.incomeAndLiabilityAccountOptions = scope.incomeAccountOptions.concat(scope.liabilityAccountOptions);
```
Comment (ShruthiRajaram) [inline app/scripts/controllers/product/CreateRateController.js]: remove commented code

File: app/scripts/controllers/product/CreateRateController.js [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+            // });
+
+            //Right now only loan is accepted for a rate.
+            scope.rateOptions = [{id : "m_loan"}];
+
+
+            scope.rateAppliesToSelected = function (rateAppliesId) {
```
Comment (ShruthiRajaram) [inline app/scripts/controllers/product/CreateRateController.js]: what is the purpose of this function

File: app/scripts/controllers/product/CreateRateController.js [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+                    scope.choice = 0;
+                }
+            };
+
+            scope.submit = function () {
+                this.formData.locale = scope.optlang.code;
+                resourceFactory.rateResource.save(this.formData, function (data) {
+                    location.path('/rates/');
+                },function(error){
+                    console.log(error);
```
Comment (ShruthiRajaram) [inline app/scripts/controllers/product/CreateRateController.js]: remove console log

File: app/scripts/controllers/product/EditLoanProductController.js [hunk 8]
```diff
@@ -241,8 +245,71 @@
...
+                  scope.rateFlag=false;
+                }
+                scope.formData.minInterestRatePerPeriod = minRate;
+
+                //Assign the same total range to this values.
+                scope.formData.interestRatePerPeriod = total;
+                scope.formData.maxInterestRatePerPeriod = total;
+                scope.calculatedRatePerPeriod = total;
+
+                console.log(scope.formData);
```
Comment (ShruthiRajaram) [inline app/scripts/controllers/product/EditLoanProductController.js:305]: remove console log

Comment (ShruthiRajaram) [reply app/scripts/controllers/product/EditLoanProductController.js:305]: also remove extra new lines  used in different files

**Review by ShruthiRajaram — CHANGES_REQUESTED** [review]

File: app/scripts/controllers/product/CreateRateController.js (hunk 5 above)
Comment (angelboxes) [reply app/scripts/controllers/product/CreateRateController.js]: Ok

File: app/scripts/controllers/product/CreateLoanProductController.js (hunk 4 above)
Comment (angelboxes) [reply app/scripts/controllers/product/CreateLoanProductController.js]: Ok

File: app/scripts/controllers/loanAccount/NewLoanAccAppController.js (hunk 3 above)
Comment (angelboxes) [reply app/scripts/controllers/loanAccount/NewLoanAccAppController.js]: Ok.

File: app/scripts/controllers/loanAccount/EditLoanAccAppController.js (hunk 1 above)
Comment (angelboxes) [reply app/scripts/controllers/loanAccount/EditLoanAccAppController.js]: Ok.

File: app/scripts/controllers/product/CreateRateController.js (hunk 6 above)
Comment (angelboxes) [reply app/scripts/controllers/product/CreateRateController.js]: Honestly it doesn't seems to do something important. Someone else worked in the code before me so my best bet is that it was used just for some tests

File: app/scripts/controllers/product/CreateRateController.js (hunk 7 above)
Comment (angelboxes) [reply app/scripts/controllers/product/CreateRateController.js]: Ok.

File: app/scripts/controllers/product/EditLoanProductController.js (hunk 8 above)
Comment (angelboxes) [reply app/scripts/controllers/product/EditLoanProductController.js:305]: Ok

File: app/scripts/controllers/loanAccount/EditLoanAccAppController.js (hunk 2 above)
Comment (angelboxes) [reply app/scripts/controllers/loanAccount/EditLoanAccAppController.js:213]: That's right and I know that we are not using strict mode, however I just like to use three equals signs whenever I can, my IDE suggest me to do it when I work in javascript files and I don't think there is something bad in using it. If you feel it should be changed then just say it.

**Comment by awasum** [comment]
@vorburger @angelboxes is this PR a complement to https://issues.apache.org/jira/browse/FINERACT-614 and https://github.com/apache/fineract/pull/592 ?

Seems there are merge conflicts, @angelboxes , will u like to adjust them?

**Comment by angelboxes** [comment]
> @vorburger @angelboxes is this PR a complement to https://issues.apache.org/jira/browse/FINERACT-614 and [apache/fineract#592](https://github.com/apache/fineract/pull/592) ?
> 
> Seems there are merge conflicts, @angelboxes , will u like to adjust them?

Hi Awasum, sorry for the late response, yes it is related, let me check the merge conflict and make sure that this work with the latest fineract version.

**Comment by edcable** [comment]
@angelboxes Are you able to check the merge conflicts here? We are FINALLY going to ship this in Fineract 1.4 and Mifos X 20.08

**Comment by angelboxes** [comment]
> @angelboxes Are you able to check the merge conflicts here? We are FINALLY going to ship this in Fineract 1.4 and Mifos X 20.08

Sorry Ed, I wasn't able to read any messages I've received from github, I will check the confilcts

**Comment by angelboxes** [comment]
Hi @edcable, I've reabased the branch twice to fix the merge conflicts.

**Comment by edcable** [comment]
Sorry @angelboxes there's one more merge conflict now. Can you please address?

**Comment by edcable** [comment]
@bharathcgowda have you had a chance to do functional review of this yet?

**Review by bharathcgowda — APPROVED** [review]
Functionality is good to have, which enhances the user experience. Functionally I would like to approve this Feature.

**Comment by awasum** [comment]
@angelboxes  Thanks for all your work. Just fix this last merge conflict on the tranlation file and I will merge this PR immediately after that. Again, sorry for the delay.

**Review by awasum — APPROVED** [review]
LGTM except the small merge conflict with the translation file. I will merge this once that is fixed.

**Comment by angelboxes** [comment]
Hi everyone, I've update again my branch with the latest changes from dev.

**Pull request merged**